Add join/leave notifications for private streams

hamatilo opened the pull request:
Implement fixes for #2746, since the issue seems to still be open. Changes were made to actions.py and streams.py for the user add and leave notifications, respectively. I also changed a few test cases in test_subs.py to account for the new notifications. In regard to the increased query count, my solution to the new notification required a user profile object when all that was accessible was a string of the user's name, which is why I increased the query count in the test cases. The query that I've implemented was an algorithm I referenced from elsewhere in the actions.py file.


Testing plan:
Ran the linter and tests provided by past contributors.



GIFs or Screenshots: As seen in the pictures, there are different messages for 1, 2, or >3 people being added and also the user left message.
![image](https://user-images.githubusercontent.com/37418212/115974708-81d07b00-a52c-11eb-954e-6d2b535d81f0.png)

hamatilo committed:
Add join/leave notifications for private streams -- issue #2746

hamatilo committed:
stuff for first issue

hamatilo committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

hamatilo committed:
Style fixes for issue #2746

hamatilo committed:
Fixed a bug dealing with bot users -- no message should be sent when a bot gets added to stream

hamatilo committed:
Add to test_subs for information including the new notifications

hamatilo committed:
Merge branch 'master' into master

PIG208 commented:
Thank you for your contributions! My first advice for you is to clean up your commit history. Please look at the [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-discipline) for reference. I would recommend you avoid having separate commits for "fix a bug", "style changes", "run ./tools/lint" and etc. You may find `git rebase -i` handy to squash the commits.

PIG208 commented:
The CI bug is likely just another problem with the test data. Specifically, you will need to update `message_id` in `zerver/openapi/curl_param_value_generators.py` as your change in `action.py`
```python
if len(notifications) > 0:
    do_send_messages(notifications, mark_as_read=[user_profile.id])
```
will send additional messages. I think you have figured this out yourself for other tests. In this case, I found the appropriate message_id to be 43.
Also please remember to add detailed descriptions for why you change the numbers in those tests to avoid confusion.

PIG208 commented on zerver/lib/actions.py:
```diff
@@ -3465,6 +3465,28 @@ def send_peer_subscriber_events(
+            notifications = []
+            for user_id in altered_user_ids:
+                try:
+                    user_profile = get_user_by_id_in_realm_including_cross_realm(user_id, realm)
```
This get function makes db query and we might not want to do it within a loop (as round-trips to the database slow everything down). Instead, please consider fetching all the user profiles outside the loop.

PIG208 commented on zerver/lib/actions.py:
```diff
@@ -3465,6 +3465,28 @@ def send_peer_subscriber_events(
+                if user_profile.is_bot:
+                    continue  # nocoverage
+
+                msg = "{} left {}.".format(user_profile.full_name, stream_dict[stream_id].name)
```
We don't usually use plain text for strings for reasons of localization. Please take a look at [i18n](https://zulip.readthedocs.io/en/latest/translating/internationalization.html?highlight=internalization) to support translation. I think we might want something like `_("{...} left {..}").format(...)` here.

PIG208 commented on zerver/tests/test_subs.py:
```diff
@@ -1642,7 +1642,7 @@ def test_realm_admin_remove_others_from_subbed_private_stream(self) -> None:
         """
         result = self.attempt_unsubscribe_of_principal(
-            query_count=17,
+            query_count=33,
```
Please include the reasoning for these changes to the query count in your commit message.

PIG208 commented on zerver/views/streams.py:
```diff
@@ -532,11 +532,32 @@ def add_subscriptions_backend(
+            names = " and ".join(newly_added_users)
+        else:
+            names = "{} people".format(len(newly_added_users))
+        msg = "{} added {} to {}.".format(user_profile.full_name, names, stream.name)
```
Similarly, we also want to follow our i18n convention here.

hamatilo committed:
Update more tests to include new notifications, restructure string formatting to i18n convention

style changes for test_subs.py

More changes to test_subs

Add # nocoverage comment for error checking case

Had to run ./tools/lint --fix for style reformat

Update more tests to include new notifications, restructure string formatting to i18n convention

Update more tests to include new notifications, restructure string formatting to i18n convention
